File: clients/ios/Lextures/Features/Courses/MarkdownRenderStyle.swift
import CoreGraphics
import Foundation

/// Spacing and typography tokens for the shared markdown renderer (CT.M2 FR-11 / FR-13).
enum MarkdownRenderStyle {
    static func blockSpacing(compact: Bool) -> CGFloat { compact ? 6 : 12 }

    static func headingPointSize(level: Int, compact: Bool) -> CGFloat {
        if compact {
            switch level {
            case 1: return 18
            case 2: return 16
            default: return 15
            }
        }
        switch level {
        case 1: return 24
        case 2: return 19
        default: return 16
        }
    }

    static func headingTopPadding(level: Int, compact: Bool) -> CGFloat {
        if compact { return level == 1 ? 2 : 0 }
        return level == 1 ? 6 : 2
    }

    /// FR-13: copy, table expand, and link navigation are suppressed during lockdown quizzes.
    static func allowsAffordances(suppressAffordances: Bool) -> Bool { !suppressAffordances }

    /// Strip link destinations so AttributedString does not create tappable runs (choices / lockdown).
    static func sourceForInlineRendering(_ text: String, suppressLinks: Bool) -> String {
        guard suppressLinks else { return text }
        return text.replacingOccurrences(
            of: #"\[([^\]]+)\]\([^)]*\)"#,
            with: "$1",
            options: .regularExpression
        )
    }

    static func inlineAttributedString(_ text: String, suppressLinks: Bool = false) -> AttributedString {
        let source = sourceForInlineRendering(text, suppressLinks: suppressLinks)
        return (try? AttributedString(
            markdown: source,
            options: .init(interpretedSyntax: .inlineOnlyPreservingWhitespace)
        )) ?? AttributedString(source)
    }

    /// Plain label for a11y / merged choice semantics (formatting must not fragment the name).
    static func plainLabel(_ text: String) -> String {
        AccessibilitySupport.plainText(fromMarkdown: text)
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/features/courses/markdown/MarkdownRenderStyle.kt
package com.lextures.android.features.courses.markdown

import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.lextures.android.core.accessibility.AccessibilitySupport

/** Spacing and typography tokens for the shared markdown renderer (CT.M2 FR-11 / FR-13). */
object MarkdownRenderStyle {
    fun blockSpacing(compact: Boolean): Dp = if (compact) 6.dp else 12.dp

    fun headingSize(level: Int, compact: Boolean): Int =
        when {
            compact && level == 1 -> 18
            compact && level == 2 -> 16
            compact -> 15
            level == 1 -> 24
            level == 2 -> 19
            else -> 16
        }

    fun headingTopPadding(level: Int, compact: Boolean): Dp =
        when {
            compact && level == 1 -> 2.dp
            compact -> 0.dp
            level == 1 -> 6.dp
            else -> 2.dp
        }

    fun allowsAffordances(suppressAffordances: Boolean): Boolean = !suppressAffordances

    fun plainLabel(text: String): String = AccessibilitySupport.plainTextFromMarkdown(text)
}
